dm/master/etcd.go

```diff
@@ -59,3 +68,111 @@ func startEtcd(masterCfg *Config,
 	}
 	return e, nil
 }
+
+// prepareJoinEtcd prepares config needed to join an existing cluster.
+// learn from https://github.com/pingcap/pd/blob/37efcb05f397f26c70cda8dd44acaa3061c92159/server/join/join.go#L44.
+func prepareJoinEtcd(cfg *Config) error {
+	// no need to join
+	if cfg.Join == "" {
+		return nil
+	}
+
+	// try to join self, invalid
+	if cfg.Join == cfg.AdvertisePeerUrls {
+		return terror.ErrMasterJoinEmbedEtcdFail.Generate(fmt.Sprintf("join self %s is forbidden", cfg.Join))
+	}
+
+	// join with persistent data
+	joinFP := filepath.Join(cfg.DataDir, "join")
+	if _, err := os.Stat(joinFP); !os.IsNotExist(err) {
+		s, err := ioutil.ReadFile(joinFP)
+		if err != nil {
+			return terror.ErrMasterJoinEmbedEtcdFail.Delegate(err, "read persistent join data")
+		}
+		cfg.InitialCluster = strings.TrimSpace(string(s))
+		cfg.InitialClusterState = embed.ClusterStateFlagExisting
+		return nil
+	}
+
+	// restart with previous data, no `InitialCluster` need to set
+	if isDataExist(filepath.Join(cfg.DataDir, "member")) {
```

Comment:
should we check `dri/member` firstly?

Reply:
Yeh, it's better. addressed in https://github.com/pingcap/dm/pull/350/commits/dcda7b23a1dad59bbdc8dd3d69e12d9cb89cc623.